flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSink.java

```diff
@@ -0,0 +1,120 @@
+
+import java.util.Properties;
+
+import static org.apache.flink.streaming.util.TypeUtil.toTypeInfo;
+
+/**
+ * A version-agnostic Kafka {@link StreamTableSink}.
+ *
+ * <p>The version-specific Kafka consumers need to extend this class and
+ * override {@link #createKafkaProducer(String, Properties, SerializationSchema, KafkaPartitioner)}}.
+ */
+public abstract class KafkaTableSink implements StreamTableSink<Row> {
+
+	private final String topic;
+	private final Properties properties;
+	private final SerializationSchema<Row> serializationSchema;
+	private final KafkaPartitioner<Row> partitioner;
+	private final String[] fieldNames;
+	private final TypeInformation[] fieldTypes;
+
+	/**
+	 * Creates KafkaTableSink
+	 *
+	 * @param topic                 Kafka topic to consume.
+	 * @param properties            Properties for the Kafka consumer.
+	 * @param serializationSchema   Serialization schema for emitted items
+	 * @param partitioner           Partitioner to select Kafka partition for each item
+	 * @param fieldNames            Row field names.
+	 * @param fieldTypes            Row field types.
+	 */
+	public KafkaTableSink(
+			String topic,
+			Properties properties,
+			SerializationSchema<Row> serializationSchema,
+			KafkaPartitioner<Row> partitioner,
+			String[] fieldNames,
+			Class<?>[] fieldTypes) {
+		this(topic, properties, serializationSchema, partitioner, fieldNames, toTypeInfo(fieldTypes));
+	}
+
+	/**
+	 * Creates KafkaTableSink
+	 *
+	 * @param topic                 Kafka topic to consume.
+	 * @param properties            Properties for the Kafka consumer.
+	 * @param serializationSchema   Serialization schema for emitted items
+	 * @param partitioner           Partitioner to select Kafka partition for each item
+	 * @param fieldNames            Row field names.
+	 * @param fieldTypes            Row field types.
+	 */
+	public KafkaTableSink(
+			String topic,
+			Properties properties,
+			SerializationSchema<Row> serializationSchema,
+			KafkaPartitioner<Row> partitioner,
+			String[] fieldNames,
+			TypeInformation<?>[] fieldTypes) {
+
+		this.topic = Preconditions.checkNotNull(topic, "topic");
+		this.properties = Preconditions.checkNotNull(properties, "properties");
+		this.serializationSchema = Preconditions.checkNotNull(serializationSchema, "serializationSchema");
+		this.partitioner = Preconditions.checkNotNull(partitioner, "partitioner");
+		this.fieldNames = Preconditions.checkNotNull(fieldNames, "fieldNames");
+		this.fieldTypes = Preconditions.checkNotNull(fieldTypes);
+		Preconditions.checkArgument(fieldNames.length == fieldTypes.length,
+			"Number of provided field names and types does not match.");
+	}
+
+	@Override
+	public void emitDataStream(DataStream<Row> dataStream) {
+		FlinkKafkaProducerBase<Row> kafkaProducer = createKafkaProducer(topic, properties, serializationSchema, partitioner);
+		dataStream.addSink(kafkaProducer);
+	}
+
+	abstract protected FlinkKafkaProducerBase<Row> createKafkaProducer(
```

Comment:
I would put this method after the constructor and add a Javadoc for it. In general, the visibility should be in font of the `abstract` modifier.